### PR TITLE
impl Trait diagnostic/test cleanups

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

**Comment by oli-obk** [comment]
O_o `--bless` seems to be exposing local paths?

**Comment by oli-obk** [comment]
Ah no. This is because it's a path into the libstd.

**Comment by oli-obk** [comment]
I adjusted compiletest to also normalize paths into the libstd/libcore

File: src/test/ui/impl-trait/impl-generic-mismatch.stderr [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  --> $DIR/impl-generic-mismatch.rs:27:23
+   |
+LL |     fn bar<U: Debug>(&self, _: &U);
+   |            - declaration in trait here
+...
+LL |     fn bar(&self, _: &impl Debug) { }
+   |                       ^^^^^^^^^^ expected generic parameter, found `impl Trait`
+help: try changing the `impl Trait` argument to a generic parameter
+   |
+LL |     fn bar<U: Debug><U: Debug>(&self, _: &U);
```
Comment (estebank) [inline src/test/ui/impl-trait/impl-generic-mismatch.stderr]: Incorrect span for the suggestion. This should be in the `bar` for the `impl Bar for ()`

**Review by estebank — COMMENTED** [review]
Love it!

File: src/librustc_typeck/check/compare_method.rs [hunk 2]
```diff
@@ -751,8 +751,131 @@ fn compare_synthetic_generics<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
+
+                        let impl_m = tcx.hir.as_local_node_id(impl_m.def_id)?;
+                        let impl_m = tcx.hir.impl_item(hir::ImplItemId { node_id: impl_m });
+
+                        // in case there are no generics, take the spot between the function name
+                        // and the opening paren of the argument list
+                        let new_generics_span = tcx
+                            .sess
+                            .codemap()
+                            .generate_fn_name_span(impl_m.span)?
```
Comment (estebank) [inline src/librustc_typeck/check/compare_method.rs]: Use `impl_span` here instead and the output will be correct.

**Comment by estebank** [comment]
@bors r+

@oli-obk approving even though there's an incorrect suggestion. I can fix it as soon as it gets merged, but if you'll have the time to do it yourself, go ahead and stop bors from merging.

**Comment by oli-obk** [comment]
I was debugging this yesterday, but didn't get very far. I think the suggestion is correct, the issue is the renderer. We should probably make it a rustfix test so we can see the fixed output

**Pull request merged**